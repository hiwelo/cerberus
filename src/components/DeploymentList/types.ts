import { DeploymentConnection } from '../../types/GitHubApi/Deployment';

export interface DeploymentListProps {
  /** List of deployments to display */
  list: DeploymentConnection;
}
