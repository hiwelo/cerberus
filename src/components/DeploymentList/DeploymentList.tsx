import React, { FunctionComponent } from 'react';
import { DeploymentCard } from 'components';
import { DeploymentListProps } from './types';

export const DeploymentList: FunctionComponent<DeploymentListProps> = ({
  list,
}) => {
  return (
    <ul>
      {list.edges.map(deployment => (
        <li key={deployment.node.commitOid}>
          <DeploymentCard item={deployment.node} />
        </li>
      ))}
    </ul>
  );
};
