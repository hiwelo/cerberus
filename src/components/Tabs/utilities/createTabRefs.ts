import { useRef } from 'react';
import { TabsProps } from '../types';

/**
 * Creates a series of React Ref for each panel part of the Tabs defined for
 * this `Tabs` component
 *
 * @param tabs The dictionary of Tabs for the related Tabs component
 */
export function createTabRefs(tabs: TabsProps['tabs']): TabsProps['tabs'] {
  // Transforms the TabListDictionary as an array of tabs
  const tabsEntries = Object.entries(tabs);
  const tabsWithRefs: TabsProps['tabs'] = {};

  // Loops over each tab and creates a React Ref for each panel HTMLElement
  tabsEntries.forEach(([tabKey, tabInfo]) => {
    const refPanel = useRef<HTMLDivElement>(null);

    tabsWithRefs[tabKey] = { ...tabInfo, refPanel };
  });

  return tabsWithRefs;
}
