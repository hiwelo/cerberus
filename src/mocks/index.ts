export { mockedActor } from './Actor';
export { mockedCommit } from './Commit';
export { mockedDeployment } from './Deployment';
export { mockedPullRequest } from './PullRequest';
