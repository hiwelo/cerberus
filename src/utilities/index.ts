export { breakpoints } from './breakpoints';
export { listNodesFromConnection } from './listNodesFromConnection';
