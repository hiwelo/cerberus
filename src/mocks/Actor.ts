import { Actor } from 'types/GitHubApi/Actor';

/**
 * Mocked Actor object
 */
export const mockedActor: Actor = {
  avatarUrl: 'https://i.pravatar.cc/300',
  login: 'test',
  url: 'https://github.com/test',
};
