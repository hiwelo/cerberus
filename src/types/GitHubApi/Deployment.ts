import { Actor } from './Actor';
import { Commit } from './Commit';
import { DateTime, URI } from './GitHubApi';

/** The possible states in which a deployment can be. */
type DeploymentState =
  | 'ABANDONED' // The pending deployment was not updated after 30 minutes
  | 'ACTIVE' // The deployment is currently active
  | 'DESTROYED' // An inactive transient deployment
  | 'ERROR' // The deployment experienced an error
  | 'FAILURE' // The deployment has failed
  | 'INACTIVE' // The deployment is inactive
  | 'PENDING' // The deployment is pending
  | 'QUEUED' // The deployment has queued
  | 'IN_PROGRESS'; // The deployment is in progress

/** Describes the status of a given deployment attempt */
type DeploymentStatus = {
  /** Identifies the actor who triggered the deployment */
  creator: Actor;
  /** Identifies the description of the deployment */
  description: string;
  /** Identifies the environment URL of the deployment */
  environmentUrl: URI;
  /** Identifies the current state of the deployment */
  state: DeploymentStatusState;
  /** Identifies the date and time when the object was last updated */
  updatedAt: DateTime;
};

/** The possible states for a deployment status */
type DeploymentStatusState =
  | 'PENDING' // The deployment is pending
  | 'SUCCESS' // The deployment was successful
  | 'FAILURE' // The deployment has failed
  | 'INACTIVE' // The deployment is inactive
  | 'ERROR' // The deployment experienced an error
  | 'QUEUED' // The deployment is queued
  | 'IN_PROGRESS'; // The deployment is in progress

/** Represents triggered deployment instance */
export interface Deployment {
  /** Identifies the oid of the deployment commit, even if the commit has been deleted */
  commitOid: string;
  /** Identifies the actor who triggered the deployment */
  creator: Actor;
  /** The deployment description */
  description: string;
  /** The environment to which this deployment was made */
  environment: string;
  /** The latest status of this deployment */
  latestStatus: DeploymentStatus;
  /** The current state of the deployment */
  state: DeploymentState;
  /** The deployment task */
  task: string;
  /** Identifies the date and time when the object was last updated */
  updatedAt: DateTime;
  /** Identifies the commit sha of the deployment */
  commit?: Commit;
}

/** The connection type for Deployment */
export interface DeploymentConnection {
  edges: DeploymentEdge[];
}

/** A deployment node in a connection */
export interface DeploymentEdge {
  __typename: 'DeploymentEdge';
  node: Deployment;
}
