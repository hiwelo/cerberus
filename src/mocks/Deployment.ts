import { Deployment, DeploymentConnection } from 'types/GitHubApi/Deployment';
import { mockedActor } from './Actor';
import { mockedCommit } from './Commit';

export const mockedDeployment: Deployment = {
  commit: mockedCommit,
  commitOid: '7638417db6d59f3c431d3e1f261cc637155684cd',
  creator: mockedActor,
  description: 'Test deployment',
  environment: 'test',
  latestStatus: {
    creator: mockedActor,
    description: 'Test environment',
    environmentUrl: 'https://example.com/',
    state: 'SUCCESS',
    updatedAt: '2019-12-20T20:11:45+00:00',
  },
  state: 'ACTIVE',
  task: 'deploy',
  updatedAt: '2019-12-20T20:11:45+00:00',
};

export const mockedDeploymentConnection: DeploymentConnection = {
  edges: [
    {
      __typename: 'DeploymentEdge',
      node: mockedDeployment,
    },
  ],
};
