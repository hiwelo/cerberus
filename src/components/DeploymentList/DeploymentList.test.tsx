import React from 'react';
import { mockedDeploymentConnection } from 'mocks/Deployment';
import { testA11y, testSnapshot } from 'utilities/test';
import { DeploymentList } from './DeploymentList';

const TestContainer = <DeploymentList list={mockedDeploymentConnection} />;

describe('DeploymentList', () => {
  testSnapshot(TestContainer);
  testA11y(TestContainer);
});
