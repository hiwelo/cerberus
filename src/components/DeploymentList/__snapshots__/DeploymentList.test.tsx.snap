// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeploymentList should match snapshot 1`] = `
<ul>
  <li>
    <span>
      deploy
    </span>
  </li>
</ul>
`;
